util/cpuprofile/cpuprofile_test.go

```diff
@@ -0,0 +1,255 @@
+	GlobalCPUProfiler.Start()
+	defer GlobalCPUProfiler.Close()
+
+	// Test register/unregister nil
+	Register(nil)
+	require.Equal(t, 0, GlobalCPUProfiler.ConsumersCount())
+	Unregister(nil)
+	require.Equal(t, 0, GlobalCPUProfiler.ConsumersCount())
+	require.False(t, GlobalCPUProfiler.inProfilingStatus())
+
+	// Test profile error and duplicate register.
+	dataCh := make(ProfileConsumer, 10)
+	err := pprof.StartCPUProfile(bytes.NewBuffer(nil))
+	require.NoError(t, err)
+
+	// Test for duplicate register.
+	Register(dataCh)
+	Register(dataCh)
+	require.Equal(t, 1, GlobalCPUProfiler.ConsumersCount())
+	require.Equal(t, true, GlobalCPUProfiler.needProfile())
+
+	// Test profile error
+	data := <-dataCh
+	require.Equal(t, "cpu profiling already in use", data.Error.Error())
+	Unregister(dataCh)
+	require.Equal(t, 0, GlobalCPUProfiler.ConsumersCount())
+
+	// shouldn't receive data from a unregistered consumer.
+	data = nil
+	select {
+	case data = <-dataCh:
+	default:
+	}
+	require.Nil(t, data)
+
+	// unregister not exist consumer
+	Unregister(dataCh)
+	require.Equal(t, 0, GlobalCPUProfiler.ConsumersCount())
+
+	// Test register a closed consumer
+	dataCh = make(ProfileConsumer, 10)
+	close(dataCh)
+	Register(dataCh)
+	require.Equal(t, 1, GlobalCPUProfiler.ConsumersCount())
+	data, ok := <-dataCh
+	require.Nil(t, data)
+	require.False(t, ok)
+	Unregister(dataCh)
+	require.Equal(t, 0, GlobalCPUProfiler.ConsumersCount())
+	pprof.StopCPUProfile()
+
+	// Test successfully get profile data.
+	dataCh = make(ProfileConsumer, 10)
+	Register(dataCh)
+	data = <-dataCh
+	require.NoError(t, data.Error)
+	profileData, err := profile.ParseData(data.Data.Bytes())
+	require.NoError(t, err)
+	require.NotNil(t, profileData)
+	Unregister(dataCh)
+	require.Equal(t, 0, GlobalCPUProfiler.ConsumersCount())
+
+	// Test stop profiling when no consumer.
+	Register(dataCh)
+	for {
+		// wait for ParallelCPUProfiler do profiling successfully
+		err = pprof.StartCPUProfile(bytes.NewBuffer(nil))
+		if err != nil {
+			break
+		}
+		pprof.StopCPUProfile()
+		time.Sleep(time.Millisecond)
+	}
+	Unregister(dataCh)
+	require.Equal(t, 0, GlobalCPUProfiler.ConsumersCount())
+
+	// wait for ParallelCPUProfiler stop profiling
+	start := time.Now()
+	for {
+		err = pprof.StartCPUProfile(bytes.NewBuffer(nil))
+		if err == nil || time.Since(start) >= time.Second*2 {
+			break
+		}
+		time.Sleep(time.Millisecond)
+	}
+	require.NoError(t, err)
+	pprof.StopCPUProfile()
+}
+
+func TestGetCPUProfile(t *testing.T) {
+	if GlobalCPUProfiler != nil {
+		GlobalCPUProfiler.Close()
+	}
+	GlobalCPUProfiler = NewParallelCPUProfiler()
+	GlobalCPUProfiler.Start()
+	defer GlobalCPUProfiler.Close()
+
+	// Test profile error
+	err := pprof.StartCPUProfile(bytes.NewBuffer(nil))
+	require.NoError(t, err)
+	err = GetCPUProfile(context.Background(), 1, bytes.NewBuffer(nil))
+	require.Error(t, err)
+	require.Equal(t, "cpu profiling already in use", err.Error())
+	pprof.StopCPUProfile()
```

Comment:
Why do we need to test it failed, intentionally? Do you think someone will call `pprof.StartCPUProfile` directly?

Reply:
The test was just used to improve the test coverage.